### ISSUE TITLE
Improve SSR-Auth functionality to allow authenticated calls from the server to Firebase

### ISSUE THREAD

**Issue by lupas** [nuxt-community/firebase-module#117]
Currently, the [SSR-Auth functionality](https://firebase.nuxtjs.org/guide/options/#auth) allows the following:

- [x] Verify the user on server-side and return the authUser object from Firebase Auth on the first call.

This way we can e.g. server-side render authenticated routes on the first call, or render the display name or email-address.

What we currently cannot do is:

- [ ] Server-side fetching of data that needs authentication from other Firebase services

For that, we would need to extend the current functionality to "sign in" the Firebase instance on server-side. Also discussed in https://github.com/nuxt-community/firebase-module/issues/69.

Seems to be a valid request, I will try to find a solution for this.

**Comment by dword-design** [comment]
@lupas This would actually be possible if the server would use the REST API instead of the client SDK because in this case the id token could be passed as a GET parameter in each request. The downside is that the interface differs between client and server. I built an (experimental) package myself that allows to add Firestore asyncData to a page and then internally it takes the .path property from each ref and calls the REST API. Not optimal but better than nothing.

**Comment by lupas** [comment]
@dword-design
Interesting, I assume that should work. Would love to have a look at your experimental package!

But as you said, writing two sets of code for SSR and Client is quite a big downside and kind of defeats the purpose and gains coming from Nuxt and the Firebase SDK.

Do you think there's a way to authenticate the Firebase instance on server-side instead? I need to take a closer look, but if that's possible that would optimal.

**Comment by dword-design** [comment]
@lupas Sure, there you go: https://github.com/dword-design/nuxt-firebase
It basically does stuff similar to your package but auths via cookies instead of service worker and adds a function firestoreData which you can use to setup asyncData for the server and client reactive data in one go. I added a short readme.

**Comment by dword-design** [comment]
Ah and regarding signing into the client SDK server-side: The problem is that the client SDK keeps the session after being logged in. On the server this doesn't work when having multiple users. At least I don't see any possibility to do this.

**Comment by lupas** [comment]
> @lupas Sure, there you go: https://github.com/dword-design/nuxt-firebase
> It basically does stuff similar to your package but auths via cookies instead of service worker and adds a function firestoreData which you can use to setup asyncData for the server and client reactive data in one go. I added a short readme.

Thanks, will have a look at it! I'll let you know what I think.

> Ah and regarding signing into the client SDK server-side: The problem is that the client SDK keeps the session after being logged in. On the server this doesn't work when having multiple users. At least I don't see any possibility to do this.

Yes, that's the thing I'm concerned about, too. Will do some investigating, but I'm not sure if there's a way with the current version of the client SDK.

**Comment by mentAl-maZe** [comment]
Hi @lupas 

There actually is a way to sign in the firebase client sdk serverside.
However it comes with a MAJOR caveat.

Firebase authentication attepts are rate limited by IP.
If an application makes heavy use of an API provided by a universal nuxt server app it will run into a 'too many authentication attempts' error.

I use your plugin in a project of mine and just need the authentication for SSR (no api calls) therefore i currently use a fork installed directly from github.

In that fork i updated the types and got rid of the serverMiddleware bug #87 by directly generating a middleware function in lieu of using a template.

Further i augmented the `firebase.services.auth.initialize.ssr` property to support the following configuration in addition to a simple boolean true:
```js
firebase.services.auth.initialize.ssr = boolean | {
  credential?: string
  serverLogin?: boolean
}
```
- `credential`:  
  takes either a nuxt.resolver path (e.g. '~/<path>')  
  or a googleServiceAccount object

  If this is omitted the module falls back on using the `admin.credential.applicationDefault()`
- `serverLogin`:  
  Server login is disabled by default to avoid API issues (as explained above)

I created a pull request so you can give it a look (#121)
As always feel free to reject or request adjustments :grin: 

PS: Sry that i never got back to you on issue #69 
I was away over new years and had a lot of catching up to do when i came back :sweat_smile:

**Comment by lupas** [comment]
> I was away over new years and had a lot of catching up to do when i came back 😅

And now back home under virus lockdown nothing better to do than coding, right? ;-) No worries, I had busy weeks myself. Wanna push this module a bit further the next few days, so I'm super glad for your help!

The rest we discuss directly in the PR!

**Comment by mentAl-maZe** [comment]
> And now back home under virus lockdown nothing better to do than coding, right? ;-)

Preach it brother :rofl:

**Comment by lupas** [comment]
Implemented an **experimental** version where serverLogin seems to work. 🎉 

> Ah and regarding signing into the client SDK server-side: The problem is that the client SDK keeps the session after being logged in. On the server this doesn't work when having multiple users. At least I don't see any possibility to do this.

@dword-design 
Have a look at our [approach in v5](https://firebase.nuxtjs.org/guide/options/#server-side-firebase-client-sdk-login), we create a separate firebase app/session for every authenticated user to avoid authorization context leakage. @mentAl-maZe did quite some magic there. 🙏 Still experimental though, have a look at it!

**Comment by dword-design** [comment]
@lupas Sorry was busy the last days. Look pretty cool. Do I understand it correctly that it is an auth login per request? And that's why there are the rate limit issues?

**Comment by mentAl-maZe** [comment]
@dword-design 
Yes and no.

Yes, every request has to be authorised, therefore stumbling into rate limiting issues on heavy traffic sites.

No, to reduce rate limiting issues we have implemented a user session lifetime.
If implemented correctly (Nuxt ist not rebuilt on every request) firebase keeps state on the server, thus forcing us to delete the created instance on each `renderRoute:done`.
As we have to check this anyways we implemented the possibility to keep the firebase instance for a certain amount of time, thus reducing the number of logins needed.

This mitigates the rate limiting issue but does not avoid it completely:
This feature is useful in use cases where an app has to perform multiple calls to the backend which make use of firebase client services, not in use cases where multiple users access the backend at the same time.
